pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+        return _abbrev_program(self.program) + '\n' + self.experiments_string(abbrev_after=20)
+
+    def serializable(self):
+        return {
+            'type': 'ExperimentSuite',
+            'experiments': self._experiments,
+            'program': self.program.out(),
+            'qubits': self.qubits,
+        }
+
+    def __eq__(self, other):
+        if not isinstance(other, ExperimentSuite):
+            return False
+        return self.serializable() == other.serializable()
+
+
+class OperatorEncoder(JSONEncoder):
+    def default(self, o):
+        if isinstance(o, Experiment):
+            return str(o)
+        if isinstance(o, ExperimentSuite):
+            return {
+                'type': 'ExperimentSuite',
+                'experiments': o._experiments,
+                'program': o.program.out(),
+                'qubits': o.qubits,
+            }
+        if isinstance(o, ExperimentResult):
+            return {
+                'type': 'ExperimentResult',
+                'experiment': o.experiment,
+                'expectation': o.expectation,
+                'stddev': o.stddev,
+            }
+
+        return o
+
+
+def to_json(fn, obj):
+    with open(fn, 'w') as f:
+        json.dump(obj, f, cls=OperatorEncoder, indent=2)
+    return fn
+
+
+def _operator_object_hook(obj):
+    if 'type' in obj and obj['type'] == 'ExperimentSuite':
+        return ExperimentSuite([[Experiment.from_str(e) for e in exps]
+                                for exps in obj['experiments']],
+                               program=Program(obj['program']),
+                               qubits=obj['qubits'])
+    return obj
+
+
+def read_json(fn):
+    with open(fn) as f:
+        return json.load(f, object_hook=_operator_object_hook)
+
+
+def _local_pauli_eig_prep(op: str, idx: int):
+    """
+    Generate gate sequence to prepare a the +1 eigenstate of a Pauli operator, assuming
+    we are starting from the ground state (the +1 eigenstate of Z^{\\otimes n}).
+
+    :param op: A string representation of the Pauli operator whose eigenstate we'd like to prepare.
+    :param idx: The index of the qubit that the preparation is acting on
+    :return: A program which will prepare the requested state.
+    """
+    if op == 'X':
+        gate = RY(pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(-pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+
+    return Program(gate)
+
+
+def _local_pauli_eig_meas(op, idx):
+    """
+    Generate gate sequence to measure in the eigenbasis of a Pauli operator, assuming
+    we are only able to measure in the Z eigenbasis.
+
+    """
+    if op == 'X':
+        gate = RY(-pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+    return Program(gate)
+
+
+def _ops_belong_to_a_tpb(op_code1: str, op_code2: str):
```

Comment:
This function could be safer if we added an `assert` statement that both the inputs are acceptable operators. Right now, `_ops_belong_to_a_tpb('nonsense', 'nonsense')` would be `True`.

Reply:
will do. in my defense this is a private helper function that is used only in the function below, but you never know how it will be used in the future